### ISSUE TITLE
Gradient Regularizer

### ISSUE THREAD

**Issue by tychovdo** [voxelmorph/voxelmorph#141]
Dear Dalca, (and other authors/contributors)

First of all, thank you for your contributions to DL-based registration methods. I really enjoyed multiple of your papers and especially like the fact that the methods are provided in a clear variational framework.

I am currently trying to understand the Voxelmorph paper and have a question about some details in your implementation. If I understand correctly, the zero-mean normal prior on the velocity field z results in a KL divergence which contains a L2 norm on the vectors in the field. Intuitively this makes sense to me, as you thereby directly penalize (possibly unnecessarily) warping of voxels.

However, in your code it seems that this penalty is not applied on the field directly, but on its approximated gradients by first substracting the same field with a small offset:
https://github.com/voxelmorph/voxelmorph/blob/7f289b3613bdbccfdb4c1eff3fbff29710a65680/pytorch/losses.py#L14-L16

I am wondering how this 'gradient_loss' is derived and why this might be better than applying the L2 norm on the velocity field directly?

To me, penalizing larger values/norms in the velocity field seems a desirable property which you seem to lose in this 'gradient_loss'.

Looking forward to your reply.

Best regards,
Tycho van der Ouderaa

**Comment by adalca** [comment]
Hi @tychovdo ,

The code you are reading reflects the TMI paper, where we penalize the gardient of the displacement. Since our main implementation is in tensorflow, the pytorch version is a bit behind, and I don't think the MICCAI/MedIA version (based on variational ELBO) is impelmented in pytorch yet. @balakg can you verify this?

To understand more about the loss you mention, please see the TMI paper (https://arxiv.org/abs/1809.05231) and let me know if something is unclear.

Thank you

**Comment by tychovdo** [comment]
@adalca , thank you. That explains a lot!
